engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/source/DelayedKafkaAvroSourceFactory.scala

```diff
@@ -0,0 +1,89 @@
+  extends KafkaAvroSourceFactory[K, V](schemaRegistryProvider, processObjectDependencies, timestampAssigner)
+    with BaseKafkaDelayedSourceFactory {
+
+  override def paramsDeterminedAfterSchema: List[Parameter] = super.paramsDeterminedAfterSchema ++ List(
+    TimestampFieldParameter, DelayParameter
+  )
+
+  override protected def nextSteps(context: ValidationContext, dependencies: List[NodeDependencyValue])
+                                  (implicit nodeId: ProcessCompilationError.NodeId): NodeTransformationDefinition = {
+    case step@TransformationStep(
+      (`topicParamName`, DefinedEagerParameter(topic: String, _)) ::
+        (SchemaVersionParamName, DefinedEagerParameter(version: String, _)) ::
+        (TimestampFieldParamName, DefinedEagerParameter(field, _)) ::
+        (DelayParameterName, DefinedEagerParameter(delay, _)) :: Nil, _) =>
+      val preparedTopic = prepareTopic(topic)
+      val versionOption = parseVersionOption(version)
+      val valueValidationResult = determineSchemaAndType(prepareValueSchemaDeterminer(preparedTopic, versionOption), Some(SchemaVersionParamName))
+
+      valueValidationResult match {
+        case Valid((valueRuntimeSchema, typingResult)) =>
+          val delayValidationErrors = Option(delay.asInstanceOf[java.lang.Long]).map(d => validateDelay(d)).getOrElse(Nil)
+          val timestampValidationErrors = Option(field.asInstanceOf[String]).map(f => validateTimestampField(f, typingResult)).getOrElse(Nil)
+          val errors = delayValidationErrors ++ timestampValidationErrors
+          prepareSourceFinalResults(preparedTopic, valueValidationResult, context, dependencies, step.parameters, errors)
+        case Invalid(exc) =>
+          prepareSourceFinalErrors(context, dependencies, step.parameters, List(exc))
+      }
+    case step@TransformationStep((`topicParamName`, _) :: (SchemaVersionParamName, _) :: (TimestampFieldParamName, _) :: (DelayParameterName, _) :: Nil, _) =>
+      prepareSourceFinalErrors(context, dependencies, step.parameters, errors = Nil)
+  }
+
+  override protected def createSource(params: Map[String, Any],
+                                      dependencies: List[NodeDependencyValue],
+                                      finalState: Option[State],
+                                      preparedTopics: List[PreparedKafkaTopic],
+                                      kafkaConfig: KafkaConfig,
+                                      deserializationSchema: KafkaDeserializationSchema[ConsumerRecord[K, V]],
+                                      timestampAssigner: Option[TimestampWatermarkHandler[ConsumerRecord[K, V]]],
+                                      formatter: RecordFormatter,
+                                      flinkContextInitializer: FlinkContextInitializer[ConsumerRecord[K, V]]): KafkaSource[ConsumerRecord[K, V]] = {
+    extractDelayInMillis(params) match {
+      case millis if millis > 0 =>
```

Comment:
Does it make sense for `DelayedSource` to enable configuring delay value equal to 0 ? Since it's delayed doesn't it imply that delay should be > 0 ?

Reply:
Delay is optional, hence additional check.